### ISSUE TITLE
Migration problem

### ISSUE THREAD

**Issue by pshober** [dtamayo/reboundx#35]
Hello! I am currently having trouble in python modifying the tau_a values for many particles at the same time. I figured out that whenever I alter the tau_a value for 10 or more particles then I get the following warning: 

> RuntimeWarning: At least 10 predictor corrector loops in IAS15 did not converge. This is typically an indication of the timestep being too large.

. If this happens the integration seems to completely shut down no matter how small the timestep is set. Also, when this warning is thrown the adaptive time stepping does not work. Everything only works if I modify 9 or less, but I need to be able to do this for thousands of particles if possible?

**Comment by hannorein** [comment]
Hi Patrick, 

The difference between 9 and 10 particles seems random to me. To give you any useful feedback, I need more information and ideally a short piece of code demonstrating the effect (emphasis on _short_). Do you use REBOUNDx for the migration forces? Is it possible that your particles have close encounters? If so that could lead to all kind of problems (with different solutions).

Hanno

**Comment by dtamayo** [comment]
The way the orbit modification forces work is that they're instantaneous forces that *average* over a Keplerian orbit to damp the semimajor axis/eccentricity etc. One problem you can run into if your particles are having close encounters is that the orbit averaged effects are now not what you expected, and that can lead to more instabilities. The warning you're getting is typically from having very close encounters.
Like Hanno says, a short code snippet would help!

**Comment by pshober** [comment]
Thank you for the fast reply! I have put an abridged version of my script below as requested with comments in places where the code wasn't necessary, etc. Yes, I am trying to use REBOUNDx to apply a semi-major axis drift to particles that I generate from observational data. The particles are generated close to the Earth (<100km) because they are based on meteor/fireball observations. Based on your comments, I now feel like this is most likely the issue causing my trouble. 

```
# Setup simulation
sim = rebound.Simulation()

# Add planets
sim.add("Sun")
sim.add("399") # Earth
sim.add("301") # Moon
sim.add("Jupiter")

# Define the number of active particles
active_particles = sim.N
sim.N_active = active_particles

# Add n_particles within errors of our data
sim.add(...) 
sim.move_to_com()

rebx = reboundx.Extras(sim)
mod_effect = rebx.add("modify_orbits_forces")

# Set tau values
for j in range(n_particles):
    sim.particles[j+active_particles].params["tau_a"] = tau_values[j]

# Integrate
x = np.zeros(shape=(n_outputs,sim.N))
y = np.zeros(shape=(n_outputs,sim.N))
times = np.linspace(0., -tmax*2.*np.pi, n_outputs)
for i,time in enumerate(times):
    for j in range(sim.N):
        # Save stuff
    sim.integrate(time)
```

**Comment by dtamayo** [comment]
Looking at this a little bit closer I think this is just a tricky numerical problem for which you'd need to make a couple modifications based on what your science interests are.

I don't think it's a problem with the tau_a prescription (which is just a drag force along the particle's velocity, which presumably is what you want for atmospheric drag). I think the problem is that you're putting the particle 100 km from the Earth, in an integration that has a scale of 1 AU, so is a million times larger. So you have to difference the Earth's and the meteoroid's position to calculate the force, and they're equal to within 6 digits. IAS15 is going to have trouble with that because it's trying to get the motion right to machine precision, which is impossible when you lose 6 digits in this way.

So you'd have to think about what you want to get out. Can you get away with just integrating the Earth moon meteoroid system? Or are you trying to trace this back to NEO space? If the latter you could add a custom post timestep modification as shown in the example that moves the reference frame to the Earth after every timestep

**Comment by pshober** [comment]
So I am actually trying to apply a Yarkovsky force on the particles as I integrate them in NEO space. I have a script working that calculates the expected drift based on physical parameters in terms of change in semi-major axis over time in units of AU / Myr. After calculating this drift rate I estimate tau_a by discretizing the problem. The amount of drift is usually on the order of 10-2 to 10-5 AU/Myr for the types of objects I am modelling. 

I also use reboundx in conjunction with an atmosphere model to take into account drag like you described. I do set the units to meters and seconds in that script, which seems to works just fine.

**Comment by dtamayo** [comment]
The Yarkovsky accelerations should be no problem at all. I think the problem you're seeing when you put particles 100 km from the Earth doesn't have to do with REBOUNDx, but rather with the way the integrator IAS15 works. If you can somehow separate the two problems and only consider the Earth and Moon in the parts of the integration where you're close to the Earth that would be the easiest solution. If you need to capture the meteoroid leaving/entering the Earth's Hill sphere, then I think you need something like the frame shifts I mentioned above

**Comment by hannorein** [comment]
I agree with Dan that this seems likely to be a problem of the vastly different scales involved (100km versus 1 AU). The solutions vary depending on what you want to do in your simulation. If you just want to integrate a few days, simply use a leap-frog integrator with a fixed timestep. It might be a bit slower, but for short integration times it should not take too long and you can still get to almost machine precision. IAS15 is using one global timestep and if only two of your many particles have a close encounter, it will slow down the entire simulation. If you want to integrate for a very long time, consider using MERCURIUS which can take longer timesteps but takes into account close encounters much more efficiently. Also consider using a softening parameter to smear out the small scale gravity if that's ok with your setup (thus preventing any close encounters). You can also change the integrator after a while. I.e. run the first few days with IAS15 when you are sure no close encounters happen, then switching to MERCURIUS once particles can have encounters. 

Making lots of visualizations really helps me in those cases. Try to find out what the _physical_ process is that slows you down!

In short: you're working on a complicated problem (many particles, close encounters, additional forces) and the solution might be complicated too!

**Comment by pshober** [comment]
Thank you so much for the help Dan and Hanno! I made some modifications based on your comments and now it seems to be all working. I ended up waiting until the particles were greater than 3 hill spheres away from the Earth before changing the tau_a values and now everything seems to be all good. Thank you so much again for the fast replies!

**Comment by hannorein** [comment]
Awesome! I'm glad to hear it's working now!
